### ISSUE TITLE
Memory Leak using Crypto-Library

### ISSUE THREAD

**Issue by bstnbuck** [vlang/v#7706]
Hi, since the memory management seems to be completed, there are still memory leaks. Since my last bug report (#6095) the behavior has not changed much.

**V version:** V 0.2.1 eef73ed
**OS:**  windows, Microsoft Windows 10 Home v18363 64-bit

**What did you do?**
I have run the following code (also in Go):
--> V-Version
```v
import crypto.sha256
import rand
import time

fn main() {
	mut nonce := rand.u64()
	mut count := u64(0)
	text := 'HelloWorld'
	println('Random Nonce: ' + nonce.str())
	timestamp_start := time.now()
	println('Started: ' + timestamp_start.str())
	for {
		hashthex := sha256.hexhash(text + nonce.str())
		if hashthex.starts_with('00000') {
			println('\nHash found!: $hashthex \nText+Nonce: ' + text + nonce.str() + ' Count: ' +
				count.str() + '\n')
			break
		}
		nonce++
		count++
	}
	timestamp_stop := time.now()
	duration := timestamp_stop.unix_time_milli() - timestamp_start.unix_time_milli()
	println('Stopped: ' + timestamp_stop.str())
	println('Duration: ' + duration.str() + 'ms => ' + (duration / 1000).str() + 'sec')
	println('Hashrate: ' + (count / duration).str()+' H/ms')
}
```
Run Code with: ```v -autofree run .``` 
Output of the Code:
```
Random Nonce: 8260002893562395378
Started: 2020-12-30 10:21:23

Hash found!: 0000057f89d2fe0e1ba5aa3a0b884937ea0e1c759b94a3c84d3fd35c0cdf01fa
Text+Nonce: HelloWorld8260002893563262154 Count: 866776

Stopped: 2020-12-30 10:21:31
Duration: 8442ms => 8sec
Hashrate: 102 H/ms
```
=> Low Hashrate and high memory usage: ~ 1,5 GB!

--> Same Code in Go
```go
package main

import (
	"crypto/sha256"
	"encoding/hex"
	"math/rand"
	"strings"
	"time"
	"strconv"
	"fmt"
)

func main(){
    rand.Seed(time.Now().UnixNano())
	var nonce = int64(rand.Int63n(1000000000))
	text := "HelloWorld"
	count := int64(0)
	hash := sha256.New()
	timestampStart := time.Now()
	fmt.Println("Started: " + timestampStart.String())
	for {
		hash.Write([]byte(text + strconv.FormatInt(nonce,10)))
		hasht := hex.EncodeToString(hash.Sum(nil))
		if strings.HasPrefix(hasht, "000000") {
			fmt.Println("\nHash found!: "+hasht+" \nText+Nonce: " + text + strconv.FormatInt(nonce,10) + " Count: " + strconv.FormatInt(count,10) + "\n")
			break
		}
		hash.Reset()
		nonce++
		count++
	}
	timestampStop := time.Now()
	duration := timestampStop.Sub(timestampStart)
	fmt.Println("Stopped: " + timestampStop.String())
	fmt.Println("Duration: " + duration.String() + " => " + (duration / 1000).String())
	fmt.Println("Hashrate: " + strconv.FormatInt((count / (duration.Milliseconds())),10)+" H/ms")
}
```
Run Code with: ```go run .```
Output of the Code:
```
Started: 2020-12-30 10:26:07.2035533 +0100 CET m=+0.004000801

Hash found!: 000000009bee52258f98b2f68569793f327d2cba8f71feda71070d08e807c668
Text+Nonce: HelloWorld488954174 Count: 1424881

Stopped: 2020-12-30 10:26:08.112428 +0100 CET m=+0.912875501
Duration: 908.8747ms => 908.874µs
Hashrate: 1569 H/ms 
```
=> High Hashrate and low memory usage: ~ 30 MB

**What did you expect to see?**
Behavior as in the Go code. No memory leaks, high hashrate
 
**What did you see instead?**
Memory leaks and low hashrate.

**Comment by joe-conigliaro** [comment]
Hi @bstnbuck `in regards to -autofree`, it still does not work for all cases.
The hash rate difference between the go and v libs have a few reasons, for starters the go library has architecture specific optimised assembly block hash functions (and other std lib fns), whereas the v libs are currently just straight v (equivalent to if go was falling back to its generic implementation). Another reason could be some extra cloning going on with arrays & arguments (I cant remember off the top of my head I would need to check).

**Comment by bstnbuck** [comment]
Hi @joe-conigliaro,
thank you very much for the quick reply. Is there a work-around for the memory leaks or when can we expect them to be fixed? Is it planned that the hashrate remains as it is or will it be improved?

**Comment by joe-conigliaro** [comment]
Hi @bstnbuck, One way would be to manually insert the free's where needed. I don't have an eta on when this will be fully supported, but just know that autofree is constantly be improved to handle more cases.

My original plans regarding the crypto libs was to get a base implementation which was pure v, that way it would not rely on any C/or arch specific features meaning it could easily be compiled for multiple platforms.
Next I was planning on either:
- Using a similar approach to go and add optimised block functions for different arch's using intrinsics/assembly.
- Adding some optimised implementations which would use the underlying system crypto api's for example windows crypto api, mac crypto services, linux kernel crypto api.
We would get the benefit of the pure v version which can be compiled to run on just about anything, and also gain the speed improvements on the major platforms.

**Comment by ArtemkaKun** [comment]
**V 0.2.4**

Memory leak still here (Valgrind report)

```bash
==1127014== 
==1127014== HEAP SUMMARY:
==1127014==     in use at exit: 123,603,572 bytes in 2,788,049 blocks
==1127014==   total heap usage: 6,505,499 allocs, 3,717,450 frees, 375,947,663 bytes allocated
==1127014== 
==1127014== 12 bytes in 1 blocks are definitely lost in loss record 1 of 27
==1127014==    at 0x483E899: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==1127014==    by 0x41BCD1: malloc_noscan (in /home/yuart/Projects/v-bugs/v-bugs)
==1127014==    by 0x429596: string__plus (in /home/yuart/Projects/v-bugs/v-bugs)
==1127014==    by 0x44C840: main__main (in /home/yuart/Projects/v-bugs/v-bugs)
==1127014==    by 0x45F41C: main (in /home/yuart/Projects/v-bugs/v-bugs)
==1127014== 
==1127014== 13 bytes in 1 blocks are definitely lost in loss record 2 of 27
==1127014==    at 0x483E899: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==1127014==    by 0x41BCD1: malloc_noscan (in /home/yuart/Projects/v-bugs/v-bugs)
==1127014==    by 0x4604A5: int_str_l (in /home/yuart/Projects/v-bugs/v-bugs)
==1127014==    by 0x41F278: int_str (in /home/yuart/Projects/v-bugs/v-bugs)
==1127014==    by 0x44C274: main__main (in /home/yuart/Projects/v-bugs/v-bugs)
==1127014==    by 0x45F41C: main (in /home/yuart/Projects/v-bugs/v-bugs)
==1127014== 
==1127014== 17 bytes in 1 blocks are definitely lost in loss record 3 of 27
==1127014==    at 0x483E899: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==1127014==    by 0x41BCD1: malloc_noscan (in /home/yuart/Projects/v-bugs/v-bugs)
==1127014==    by 0x429596: string__plus (in /home/yuart/Projects/v-bugs/v-bugs)
==1127014==    by 0x44C61E: main__main (in /home/yuart/Projects/v-bugs/v-bugs)
==1127014==    by 0x45F41C: main (in /home/yuart/Projects/v-bugs/v-bugs)
==1127014== 
==1127014== 20 bytes in 1 blocks are definitely lost in loss record 4 of 27
==1127014==    at 0x483E899: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==1127014==    by 0x41BCD1: malloc_noscan (in /home/yuart/Projects/v-bugs/v-bugs)
==1127014==    by 0x41C6A7: memdup_noscan (in /home/yuart/Projects/v-bugs/v-bugs)
==1127014==    by 0x4042FD: strings__Builder_str (in /home/yuart/Projects/v-bugs/v-bugs)
==1127014==    by 0x432CE3: str_intp (in /home/yuart/Projects/v-bugs/v-bugs)
==1127014==    by 0x439E76: time__Time_get_fmt_str (in /home/yuart/Projects/v-bugs/v-bugs)
==1127014==    by 0x43529D: time__Time_format_ss (in /home/yuart/Projects/v-bugs/v-bugs)
==1127014==    by 0x43F8C4: time__Time_str (in /home/yuart/Projects/v-bugs/v-bugs)
==1127014==    by 0x44BE1B: main__main (in /home/yuart/Projects/v-bugs/v-bugs)
==1127014==    by 0x45F41C: main (in /home/yuart/Projects/v-bugs/v-bugs)
==1127014== 
==1127014== 20 bytes in 1 blocks are definitely lost in loss record 5 of 27
==1127014==    at 0x483E899: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==1127014==    by 0x41BCD1: malloc_noscan (in /home/yuart/Projects/v-bugs/v-bugs)
==1127014==    by 0x41C6A7: memdup_noscan (in /home/yuart/Projects/v-bugs/v-bugs)
==1127014==    by 0x4042FD: strings__Builder_str (in /home/yuart/Projects/v-bugs/v-bugs)
==1127014==    by 0x432CE3: str_intp (in /home/yuart/Projects/v-bugs/v-bugs)
==1127014==    by 0x439E76: time__Time_get_fmt_str (in /home/yuart/Projects/v-bugs/v-bugs)
==1127014==    by 0x43529D: time__Time_format_ss (in /home/yuart/Projects/v-bugs/v-bugs)
==1127014==    by 0x43F8C4: time__Time_str (in /home/yuart/Projects/v-bugs/v-bugs)
==1127014==    by 0x44C515: main__main (in /home/yuart/Projects/v-bugs/v-bugs)
==1127014==    by 0x45F41C: main (in /home/yuart/Projects/v-bugs/v-bugs)
==1127014== 
==1127014== 21 bytes in 1 blocks are definitely lost in loss record 6 of 27
==1127014==    at 0x483E899: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==1127014==    by 0x41BCD1: malloc_noscan (in /home/yuart/Projects/v-bugs/v-bugs)
==1127014==    by 0x41F74D: u64_str (in /home/yuart/Projects/v-bugs/v-bugs)
==1127014==    by 0x44BD12: main__main (in /home/yuart/Projects/v-bugs/v-bugs)
==1127014==    by 0x45F41C: main (in /home/yuart/Projects/v-bugs/v-bugs)
==1127014== 
==1127014== 21 bytes in 1 blocks are definitely lost in loss record 7 of 27
==1127014==    at 0x483E899: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==1127014==    by 0x41BCD1: malloc_noscan (in /home/yuart/Projects/v-bugs/v-bugs)
==1127014==    by 0x41F74D: u64_str (in /home/yuart/Projects/v-bugs/v-bugs)
==1127014==    by 0x44C1B5: main__main (in /home/yuart/Projects/v-bugs/v-bugs)
==1127014==    by 0x45F41C: main (in /home/yuart/Projects/v-bugs/v-bugs)
==1127014== 
==1127014== 21 bytes in 1 blocks are definitely lost in loss record 8 of 27
==1127014==    at 0x483E899: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==1127014==    by 0x41BCD1: malloc_noscan (in /home/yuart/Projects/v-bugs/v-bugs)
==1127014==    by 0x41F4F6: i64_str (in /home/yuart/Projects/v-bugs/v-bugs)
==1127014==    by 0x44C5DD: main__main (in /home/yuart/Projects/v-bugs/v-bugs)
==1127014==    by 0x45F41C: main (in /home/yuart/Projects/v-bugs/v-bugs)
==1127014== 
==1127014== 21 bytes in 1 blocks are definitely lost in loss record 9 of 27
==1127014==    at 0x483E899: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==1127014==    by 0x41BCD1: malloc_noscan (in /home/yuart/Projects/v-bugs/v-bugs)
==1127014==    by 0x41F4F6: i64_str (in /home/yuart/Projects/v-bugs/v-bugs)
==1127014==    by 0x44C6BD: main__main (in /home/yuart/Projects/v-bugs/v-bugs)
==1127014==    by 0x45F41C: main (in /home/yuart/Projects/v-bugs/v-bugs)
==1127014== 
==1127014== 21 bytes in 1 blocks are definitely lost in loss record 10 of 27
==1127014==    at 0x483E899: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==1127014==    by 0x41BCD1: malloc_noscan (in /home/yuart/Projects/v-bugs/v-bugs)
==1127014==    by 0x41F4F6: i64_str (in /home/yuart/Projects/v-bugs/v-bugs)
==1127014==    by 0x44C7FF: main__main (in /home/yuart/Projects/v-bugs/v-bugs)
==1127014==    by 0x45F41C: main (in /home/yuart/Projects/v-bugs/v-bugs)
==1127014== 
==1127014== 23 bytes in 1 blocks are definitely lost in loss record 11 of 27
==1127014==    at 0x483E899: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==1127014==    by 0x41BCD1: malloc_noscan (in /home/yuart/Projects/v-bugs/v-bugs)
==1127014==    by 0x429596: string__plus (in /home/yuart/Projects/v-bugs/v-bugs)
==1127014==    by 0x44C683: main__main (in /home/yuart/Projects/v-bugs/v-bugs)
==1127014==    by 0x45F41C: main (in /home/yuart/Projects/v-bugs/v-bugs)
==1127014== 
==1127014== 26 bytes in 1 blocks are definitely lost in loss record 12 of 27
==1127014==    at 0x483E899: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==1127014==    by 0x41BCD1: malloc_noscan (in /home/yuart/Projects/v-bugs/v-bugs)
==1127014==    by 0x429596: string__plus (in /home/yuart/Projects/v-bugs/v-bugs)
==1127014==    by 0x44C6FE: main__main (in /home/yuart/Projects/v-bugs/v-bugs)
==1127014==    by 0x45F41C: main (in /home/yuart/Projects/v-bugs/v-bugs)
==1127014== 
==1127014== 29 bytes in 1 blocks are possibly lost in loss record 13 of 27
==1127014==    at 0x4843A83: calloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==1127014==    by 0x41C5B8: vcalloc (in /home/yuart/Projects/v-bugs/v-bugs)
==1127014==    by 0x416312: __new_array_with_default (in /home/yuart/Projects/v-bugs/v-bugs)
==1127014==    by 0x42E73B: string_bytes (in /home/yuart/Projects/v-bugs/v-bugs)
==1127014==    by 0x448517: crypto__sha256__hexhash (in /home/yuart/Projects/v-bugs/v-bugs)
==1127014==    by 0x44BF47: main__main (in /home/yuart/Projects/v-bugs/v-bugs)
==1127014==    by 0x45F41C: main (in /home/yuart/Projects/v-bugs/v-bugs)
==1127014== 
==1127014== 32 bytes in 1 blocks are possibly lost in loss record 14 of 27
==1127014==    at 0x4843A83: calloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==1127014==    by 0x41C5B8: vcalloc (in /home/yuart/Projects/v-bugs/v-bugs)
==1127014==    by 0x416312: __new_array_with_default (in /home/yuart/Projects/v-bugs/v-bugs)
==1127014==    by 0x445D24: crypto__sha256__Digest_reset (in /home/yuart/Projects/v-bugs/v-bugs)
==1127014==    by 0x4464E8: crypto__sha256__new (in /home/yuart/Projects/v-bugs/v-bugs)
==1127014==    by 0x447F60: crypto__sha256__sum256 (in /home/yuart/Projects/v-bugs/v-bugs)
==1127014==    by 0x448554: crypto__sha256__hexhash (in /home/yuart/Projects/v-bugs/v-bugs)
==1127014==    by 0x44BF47: main__main (in /home/yuart/Projects/v-bugs/v-bugs)
==1127014==    by 0x45F41C: main (in /home/yuart/Projects/v-bugs/v-bugs)
==1127014== 
==1127014== 88 bytes in 1 blocks are possibly lost in loss record 15 of 27
==1127014==    at 0x483E899: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==1127014==    by 0x41B943: _v_malloc (in /home/yuart/Projects/v-bugs/v-bugs)
==1127014==    by 0x41C639: memdup (in /home/yuart/Projects/v-bugs/v-bugs)
==1127014==    by 0x4464D8: crypto__sha256__new (in /home/yuart/Projects/v-bugs/v-bugs)
==1127014==    by 0x447F60: crypto__sha256__sum256 (in /home/yuart/Projects/v-bugs/v-bugs)
==1127014==    by 0x448554: crypto__sha256__hexhash (in /home/yuart/Projects/v-bugs/v-bugs)
==1127014==    by 0x44BF47: main__main (in /home/yuart/Projects/v-bugs/v-bugs)
==1127014==    by 0x45F41C: main (in /home/yuart/Projects/v-bugs/v-bugs)
==1127014== 
==1127014== 93 bytes in 1 blocks are definitely lost in loss record 16 of 27
==1127014==    at 0x483E899: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==1127014==    by 0x41BCD1: malloc_noscan (in /home/yuart/Projects/v-bugs/v-bugs)
==1127014==    by 0x41C6A7: memdup_noscan (in /home/yuart/Projects/v-bugs/v-bugs)
==1127014==    by 0x4042FD: strings__Builder_str (in /home/yuart/Projects/v-bugs/v-bugs)
==1127014==    by 0x432CE3: str_intp (in /home/yuart/Projects/v-bugs/v-bugs)
==1127014==    by 0x44C160: main__main (in /home/yuart/Projects/v-bugs/v-bugs)
==1127014==    by 0x45F41C: main (in /home/yuart/Projects/v-bugs/v-bugs)
==1127014== 
==1127014== 103 bytes in 1 blocks are definitely lost in loss record 17 of 27
==1127014==    at 0x483E899: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==1127014==    by 0x41BCD1: malloc_noscan (in /home/yuart/Projects/v-bugs/v-bugs)
==1127014==    by 0x429596: string__plus (in /home/yuart/Projects/v-bugs/v-bugs)
==1127014==    by 0x44C19B: main__main (in /home/yuart/Projects/v-bugs/v-bugs)
==1127014==    by 0x45F41C: main (in /home/yuart/Projects/v-bugs/v-bugs)
==1127014== 
==1127014== 122 bytes in 1 blocks are definitely lost in loss record 18 of 27
==1127014==    at 0x483E899: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==1127014==    by 0x41BCD1: malloc_noscan (in /home/yuart/Projects/v-bugs/v-bugs)
==1127014==    by 0x429596: string__plus (in /home/yuart/Projects/v-bugs/v-bugs)
==1127014==    by 0x44C1F6: main__main (in /home/yuart/Projects/v-bugs/v-bugs)
==1127014==    by 0x45F41C: main (in /home/yuart/Projects/v-bugs/v-bugs)
==1127014==
==1127014== 128 bytes in 2 blocks are possibly lost in loss record 19 of 27
==1127014==    at 0x4843A83: calloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==1127014==    by 0x41C5B8: vcalloc (in /home/yuart/Projects/v-bugs/v-bugs)
==1127014==    by 0x416312: __new_array_with_default (in /home/yuart/Projects/v-bugs/v-bugs)
==1127014==    by 0x445DC8: crypto__sha256__Digest_reset (in /home/yuart/Projects/v-bugs/v-bugs)
==1127014==    by 0x4464E8: crypto__sha256__new (in /home/yuart/Projects/v-bugs/v-bugs)
==1127014==    by 0x447F60: crypto__sha256__sum256 (in /home/yuart/Projects/v-bugs/v-bugs)
==1127014==    by 0x448554: crypto__sha256__hexhash (in /home/yuart/Projects/v-bugs/v-bugs)
==1127014==    by 0x44BF47: main__main (in /home/yuart/Projects/v-bugs/v-bugs)
==1127014==    by 0x45F41C: main (in /home/yuart/Projects/v-bugs/v-bugs)
==1127014==
==1127014== 130 bytes in 1 blocks are definitely lost in loss record 20 of 27
==1127014==    at 0x483E899: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==1127014==    by 0x41BCD1: malloc_noscan (in /home/yuart/Projects/v-bugs/v-bugs)
==1127014==    by 0x429596: string__plus (in /home/yuart/Projects/v-bugs/v-bugs)
==1127014==    by 0x44C25B: main__main (in /home/yuart/Projects/v-bugs/v-bugs)
==1127014==    by 0x45F41C: main (in /home/yuart/Projects/v-bugs/v-bugs)
==1127014==
==1127014== 136 bytes in 1 blocks are definitely lost in loss record 21 of 27
==1127014==    at 0x483E899: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==1127014==    by 0x41BCD1: malloc_noscan (in /home/yuart/Projects/v-bugs/v-bugs)
==1127014==    by 0x429596: string__plus (in /home/yuart/Projects/v-bugs/v-bugs)
==1127014==    by 0x44C2B5: main__main (in /home/yuart/Projects/v-bugs/v-bugs)
==1127014==    by 0x45F41C: main (in /home/yuart/Projects/v-bugs/v-bugs)
==1127014==
==1127014== 9,758,112 bytes in 464,672 blocks are definitely lost in loss record 22 of 27
==1127014==    at 0x483E899: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==1127014==    by 0x41BCD1: malloc_noscan (in /home/yuart/Projects/v-bugs/v-bugs)
==1127014==    by 0x41F74D: u64_str (in /home/yuart/Projects/v-bugs/v-bugs)
==1127014==    by 0x44BEBC: main__main (in /home/yuart/Projects/v-bugs/v-bugs)
==1127014==    by 0x45F41C: main (in /home/yuart/Projects/v-bugs/v-bugs)
==1127014==
==1127014== 13,475,459 bytes in 464,671 blocks are definitely lost in loss record 23 of 27
==1127014==    at 0x4843A83: calloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==1127014==    by 0x41C5B8: vcalloc (in /home/yuart/Projects/v-bugs/v-bugs)
==1127014==    by 0x416312: __new_array_with_default (in /home/yuart/Projects/v-bugs/v-bugs)
==1127014==    by 0x42E73B: string_bytes (in /home/yuart/Projects/v-bugs/v-bugs)
==1127014==    by 0x448517: crypto__sha256__hexhash (in /home/yuart/Projects/v-bugs/v-bugs)
==1127014==    by 0x44BF47: main__main (in /home/yuart/Projects/v-bugs/v-bugs)
==1127014==    by 0x45F41C: main (in /home/yuart/Projects/v-bugs/v-bugs)
==1127014==
==1127014== 14,869,504 bytes in 464,672 blocks are definitely lost in loss record 25 of 27
==1127014==    at 0x4843A83: calloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==1127014==    by 0x41C5B8: vcalloc (in /home/yuart/Projects/v-bugs/v-bugs)
==1127014==    by 0x416312: __new_array_with_default (in /home/yuart/Projects/v-bugs/v-bugs)
==1127014==    by 0x447679: crypto__sha256__Digest_checksum (in /home/yuart/Projects/v-bugs/v-bugs)
==1127014==    by 0x4480B4: crypto__sha256__sum256 (in /home/yuart/Projects/v-bugs/v-bugs)
==1127014==    by 0x448554: crypto__sha256__hexhash (in /home/yuart/Projects/v-bugs/v-bugs)
==1127014==    by 0x44BF47: main__main (in /home/yuart/Projects/v-bugs/v-bugs)
==1127014==    by 0x45F41C: main (in /home/yuart/Projects/v-bugs/v-bugs)
==1127014==
==1127014== 85,499,400 (40,891,048 direct, 44,608,352 indirect) bytes in 464,671 blocks are definitely lost in loss record 27 of 27
==1127014==    at 0x483E899: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==1127014==    by 0x41B943: _v_malloc (in /home/yuart/Projects/v-bugs/v-bugs)
==1127014==    by 0x41C639: memdup (in /home/yuart/Projects/v-bugs/v-bugs)
==1127014==    by 0x4464D8: crypto__sha256__new (in /home/yuart/Projects/v-bugs/v-bugs)
==1127014==    by 0x447F60: crypto__sha256__sum256 (in /home/yuart/Projects/v-bugs/v-bugs)
==1127014==    by 0x448554: crypto__sha256__hexhash (in /home/yuart/Projects/v-bugs/v-bugs)
==1127014==    by 0x44BF47: main__main (in /home/yuart/Projects/v-bugs/v-bugs)
==1127014==    by 0x45F41C: main (in /home/yuart/Projects/v-bugs/v-bugs)
==1127014==
==1127014== LEAK SUMMARY:
==1127014==    definitely lost: 78,994,943 bytes in 1,858,703 blocks
==1127014==    indirectly lost: 44,608,352 bytes in 929,341 blocks
==1127014==      possibly lost: 277 bytes in 5 blocks
==1127014==    still reachable: 0 bytes in 0 blocks
==1127014==         suppressed: 0 bytes in 0 blocks
==1127014==
==1127014== For lists of detected and suppressed errors, rerun with: -s
==1127014== ERROR SUMMARY: 25 errors from 25 contexts (suppressed: 0 from 0)
```

Benchmark results

`v -autofree run .`

```bash
Random Nonce: 16945258508882931328
Started: 2022-02-12 17:46:38

Hash found!: 000005dfcad6b0c532b0cc9c0ab390a0715eb7182c93cac14f22239739772fbc
Text+Nonce: HelloWorld16945258508883331214 Count: 399886

Stopped: 2022-02-12 17:46:40
Duration: 2176ms => 2sec
Hashrate: 183 H/ms
```

`v -prod -autofree run .`

```bash
Random Nonce: 5829361815337882692
Started: 2022-02-12 17:46:54

Hash found!: 00000949cdf7e6ce2d8b337bf6597c2ff091920db1d4d42b3436552903ab1781
Text+Nonce: HelloWorld5829361815338259281 Count: 376589

Stopped: 2022-02-12 17:46:55
Duration: 258ms => 0sec
Hashrate: 1459 H/ms
```

**Comment by medvednikov** [comment]
Fixed in 0.3.
